ggml-metal.metal

```diff
@@ -1903,12 +1251,154 @@ kernel void kernel_get_rows(
 
     for (int ind = tiitg; ind < ne00/16; ind += tptg) {
         float4x4 temp;
-        dequantize_func(
-            ((device const block_q *) ((device char *) src0 + r*nb01)) + ind/nl, ind%nl, temp);
+        quant_dri<device const uint16_t *, device const block_q_type *, float4x4> dequan_worker;
+        dequan_worker.init(ind%nl);
+        dequan_worker.dequantize(
+            ((device const block_q_type *) ((device char *) src0 + r*nb01)) + ind/nl, ind%nl, temp);
         *(((device float4x4 *) ((device char *) dst + i*nb1)) + ind) = temp;
     }
 }
 
+// nl: Each block has 16*nl weights
+// n_shift: Each thread deals with 16 dequantized weights. However, the 16 weights may not be continuous.
+//          n_shift is the difference between the address of the first 8 weights and the last 8 weights.
+//          (i.e. n_shift=8 means 16 continuous weights)
+template<typename block_q_type, int nr, int nsg, int nl, int n_shift, template<typename, typename, typename> class quant_dri>
+kernel void kernel_mat_mv(device const   void * src0,
+                          device const  float * src1,
```

Comment:
This is signature is not OK - I need a few minutes of thinking every time I see it to understand what is the meaning of the template args